### PR TITLE
Bring back generic parameters for indices in rustc_abi and make it compile on stable

### PR DESCRIPTION
**Pull request by Veykril**
This effectively reverses https://github.com/rust-lang/rust/pull/107163, allowing rust-analyzer to depend on this crate again,

It also moves some glob imports / expands them in the first commit because they made it more difficult for me to reason about things.

### CODE REVIEW

File: compiler/rustc_abi/src/lib.rs [hunk 1]
```diff
@@ -1501,29 +1486,12 @@ impl Niche {
...
-        /// Equivalent to `VariantIdx(0)`.
-        const FIRST_VARIANT = 0;
-    }
-}
-
+// NOTE: This struct is generic over the FieldIdx and VariantIdx for rust-analyzer usage.
 #[derive(PartialEq, Eq, Hash, Clone)]
 #[cfg_attr(feature = "nightly", derive(HashStable_Generic))]
-pub struct LayoutS {
+pub struct LayoutS<FieldIdx: Idx, VariantIdx: Idx> {
```
Comment (WaffleLapkin) [inline compiler/rustc_abi/src/lib.rs:1492]: Could you provide defaults, such that we don't need to touch every mention of this type again? Or would this not work because the types are unavailable if this is not compiled for rustc?

**Comment by WaffleLapkin** [comment]
What's the reason for not simply reverting the commit, why did you need to re-do all changes?

**Comment by Veykril** [comment]
Because a bunch of other changes have been commited on top of it making a revert non-trivial.

File: compiler/rustc_abi/src/lib.rs [hunk 2]
```diff
@@ -28,9 +27,6 @@ pub use layout::LayoutCalculator;
 /// instead of implementing everything in `rustc_middle`.
 pub trait HashStableContext {}
 
-use Integer::*;
-use Primitive::*;
```
Comment (WaffleLapkin) [inline compiler/rustc_abi/src/lib.rs:32]: I don't mind, but just wandering, why this change?

File: compiler/rustc_abi/src/layout.rs [hunk 3]
```diff
@@ -1,21 +1,26 @@
...
-use rand_xoshiro::Xoshiro128StarStar;
-
+use rustc_index::Idx;
 use tracing::debug;
 
+use crate::{
+    Abi, AbiAndPrefAlign, Align, FieldsShape, IndexSlice, IndexVec, Integer, LayoutS, Niche,
+    NonZeroUsize, Primitive, ReprOptions, Scalar, Size, StructKind, TagEncoding, TargetDataLayout,
+    Variants, WrappingRange,
+};
```
Comment (WaffleLapkin) [inline compiler/rustc_abi/src/layout.rs:12]: Can you add one or two newlines here? Formatting looks very weird here...

**Review by WaffleLapkin — COMMENTED** [review]
Annoying that we have to do that, as the code without generics is a little easier to read, but I understand the reasoning. LGTM, modulo the formatting nitpick.

cc @mikebenfield as the author of https://github.com/rust-lang/rust/pull/107163, you might have an opinion here (is there a way to make it easier to read/make less changes, but still keep the r-a use-case?).

File: compiler/rustc_abi/src/lib.rs (hunk 2 above)
Comment (Veykril) [reply compiler/rustc_abi/src/lib.rs:32]: Since I was expanding the other globs in this module already I just moved these as well. So really just personal preference (and it makes things easier to reason about to me at least)

Comment (Veykril) [reply compiler/rustc_abi/src/lib.rs:32]: I can revert this one if wanted

Comment (WaffleLapkin) [reply compiler/rustc_abi/src/lib.rs:32]: I generally like local enum glob imports more, so no need to change anything =)

File: compiler/rustc_abi/src/layout.rs (hunk 3 above)
Comment (Veykril) [reply compiler/rustc_abi/src/layout.rs:12]: I can try but I don't think rustfmt will allow me to

**Comment by Veykril** [comment]
> is there a way to make it easier to read/make less changes, but still keep the r-a use-case?

This is as minimal as it gets as far as I can tell. We need to be generic over the index types since r-a and rustc use different ones here, and since we wan't to build on stable (and because we don't need them), all the serialize derives etc have to be gated. It would be nice if we could extract things out such that the nightly feature becomes unnecessary but given the derives for serialization and hashing are required for rustc that's not feasible.

File: compiler/rustc_abi/src/layout.rs (hunk 3 above)
Comment (Veykril) [reply compiler/rustc_abi/src/layout.rs:12]: Merged the change into the last commit

**Review by WaffleLapkin — APPROVED** [review]
